### PR TITLE
[pkg/stanza] Add an option to resend logs instead of dropping

### LINKED ISSUE #20511: [pkg/stanza] Support back-pressure from downstream consumers
Currently, we drop logs on any error from the downstream consumer. This behavior is not desired in case of temporary errors. Users might want to slow down reading the log files instead of dropping the data.

We should provide a way to have such behavior that a configuration option can control.

### PR DESCRIPTION
**Pull request by dmitryax**
Add a `retry_on_failure` config option (disabled by default) that can be used to slow down reading logs instead of dropping if downstream components return a non-permanent error. The configuration has the following options:
    - `enabled`: Enable or disable the retry mechanism. Default is `false`.
    - `initial_interval`: The initial interval to wait before retrying. Default is `1s`.
    - `max_interval`: The maximum interval to wait before retrying. Default is `30s`.
    - `max_elapsed_time`: The maximum amount of time to wait before giving up. Default is `5m`.

The configuration interface is inspired by https://github.com/open-telemetry/opentelemetry-collector/tree/main/exporter/exporterhelper#configuration 
 which potentially can be exposed in as another package not specific to exporter and used by any components

Resolves: https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/20511

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -219,8 +219,8 @@ run:
 
 .PHONY: docker-component # Not intended to be used directly
 docker-component: check-component
-	GOOS=linux GOARCH=amd64 $(MAKE) $(COMPONENT)
-	cp ./bin/$(COMPONENT)_linux_amd64 ./cmd/$(COMPONENT)/$(COMPONENT)
+	GOOS=linux GOARCH=arm64 $(MAKE) $(COMPONENT)
+	cp ./bin/$(COMPONENT)_linux_arm64 ./cmd/$(COMPONENT)/$(COMPONENT)
```
Comment (atoulme) [inline Makefile:223]: Is this change related to this PR?

Comment (dmitryax) [reply Makefile:223]: It's not :) thanks for catching

File: pkg/stanza/adapter/receiver.go [hunk 2]
```diff
@@ -132,15 +140,84 @@ func (r *receiver) consumerLoop(ctx context.Context) {
...
-			if cErr != nil {
-				r.logger.Error("ConsumeLogs() failed", zap.Error(cErr))
-			}
+			cErr := r.consumeLogs(ctx, pLogs)
 			r.obsrecv.EndLogsOp(obsrecvCtx, "stanza", pLogs.LogRecordCount(), cErr)
 		}
 	}
 }
 
+func (r *receiver) consumeLogs(ctx context.Context, logs plog.Logs) error {
```
Comment (djaglowski) [inline pkg/stanza/adapter/receiver.go:149]: Is there anything about this functionality that makes it naturally unique to the stanza package? It seems quite generic to me. Should all components exhibit this behavior the same way?

File: pkg/stanza/adapter/config.go [hunk 3]
```diff
@@ -15,13 +15,43 @@
...
 
 	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/operator"
 )
 
+// RetryConfig defines configuration for retrying batches in case of receiving a retryable error from a downstream
+// consumer. The current supported strategy is exponential backoff.
+type RetryConfig struct {
+	// Enabled indicates whether to not retry sending logs in case of receiving a retryable error from a downstream
+	// consumer. Default is false.
+	Enabled bool `mapstructure:"enabled"`
```
[the remark this answers is not in the document]
Comment (dmitryax) [reply pkg/stanza/adapter/config.go:30]: Do you suggest that users use
```yaml
filelog:
  retry_on_failure:
```
instead of
```yaml
filelog:
  retry_on_failure:
    enabled: true
```

I know we use this pattern in a couple of places like otlp and hostmetrics receiver, but I don't think it should be applied here. I'd vote for having `retry_on_failure.enabled` as an explicit option. The first option seems a bit confusing to me as a user. Also, we follow the existing interface defined in https://github.com/open-telemetry/opentelemetry-collector/tree/main/exporter/exporterhelper#configuration

File: pkg/stanza/adapter/config.go [hunk 4]
```diff
@@ -15,13 +15,43 @@
...
+	// InitialInterval the time to wait after the first failure before retrying. Default value is 1 second.
+	InitialInterval time.Duration `mapstructure:"initial_interval"`
+	// MaxInterval is the upper bound on backoff interval. Once this value is reached the delay between
+	// consecutive retries will always be `MaxInterval`. Default value is 30 seconds.
+	MaxInterval time.Duration `mapstructure:"max_interval"`
+	// MaxElapsedTime is the maximum amount of time (including retries) spent trying to send a logs batch to
+	// a downstream consumer. Once this value is reached, the data is discarded. It never stops if MaxElapsedTime == 0.
+	// Default value is 5 minutes.
+	MaxElapsedTime time.Duration `mapstructure:"max_elapsed_time"`
+}
```
Comment (dloucasfx) [inline pkg/stanza/adapter/config.go:40]: Shall we warn and document against setting `MaxElapsedTime` to `0` ; 
My concern is that we can hit a deadlock with value `0`, especially if the amount of "free memory" needed by the memory limiter, after GC, is not met (ie: keeping `mustRefuse` set to true) unless we drop the data are trying to send.

File: pkg/stanza/adapter/receiver.go (hunk 2 above)
Comment (dmitryax) [reply pkg/stanza/adapter/receiver.go:149]: Yes, I think we can expose it in https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/internal/coreinternal for now. Later we'll see if we can consolidate it with the retry functionality in core somewhere outside of the exporterhelper.

File: pkg/stanza/adapter/config.go (hunk 4 above)
Comment (dloucasfx) [reply pkg/stanza/adapter/config.go:40]: @dmitryax on a slightly related note, can we have a total max data the receiver can read/process at a time? The thought behind it, is to workaround situations where the receiver reads a large amount of logs, fills up the memory right away, and can't even send the first batch. We have seen this happening in production.

Comment (dmitryax) [reply pkg/stanza/adapter/config.go:40]: > @dmitryax on a slightly related note, can we have a total max data the receiver can read/process at a time? The thought behind it, is to workaround situations where the receiver reads a large amount of logs, fills up the memory right away, and can't even send the first batch. We have seen this happening in production.

The file reader reads logs line by line and passes each of them to the downstream operators separately, it has a `max_log_size` option to avoid reading long lines, which is 1Mb by default. Then we also have the `recombine` operator, which potentially can cause issues like that when many logs are stitched together. We recently added the same 1Mb limit in our helm chart config to avoid that. I believe that should be sufficient.

Comment (dmitryax) [reply pkg/stanza/adapter/config.go:40]: > Shall we warn and document against setting MaxElapsedTime to 0 ;
> My concern is that we can hit a deadlock with value 0, especially if the amount of "free memory" needed by the memory limiter, after GC, is not met (ie: keeping mustRefuse set to true) unless we drop the data are trying to send.

I think it's pretty unlikely that all the memory will be taken by the file log receiver. The maximum number of log records it can accumulate is 100 (the default batch size). The maximum possible size of each log is 1Mb which is very unlikely to be reached.

File: pkg/stanza/adapter/receiver.go (hunk 2 above)
Comment (djaglowski) [reply pkg/stanza/adapter/receiver.go:149]: Should we hold this PR until it's added in `coreinternal` then?

File: pkg/stanza/adapter/receiver.go [hunk 5]
```diff
@@ -132,15 +140,84 @@ func (r *receiver) consumerLoop(ctx context.Context) {
...
+		err := r.consumer.ConsumeLogs(ctx, logs)
+		if err == nil {
+			return nil
+		}
+
+		if consumererror.IsPermanent(err) {
+			r.logger.Error(
+				"ConsumeLogs() failed. The error is not retryable. Dropping data.",
+				zap.Error(err),
+				zap.Int("dropped_items", logs.LogRecordCount()),
```
Comment (sumo-drosiek) [inline pkg/stanza/adapter/receiver.go:188]: Shouldn't we extract number of dropped records from the `err` (as it should be `consumererror.Logs`)?

Comment (dmitryax) [reply pkg/stanza/adapter/receiver.go:188]: Permanent errors don't have a number of dropped records yet. Will be added in https://github.com/open-telemetry/opentelemetry-collector/pull/7439

Comment (dmitryax) [reply pkg/stanza/adapter/receiver.go]: I updated the logic to use partial retries tho.

File: pkg/stanza/adapter/receiver.go (hunk 2 above)
Comment (dmitryax) [reply pkg/stanza/adapter/receiver.go]: I moved the logic to `coreinternal` in this PR

File: pkg/stanza/adapter/config.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (dmitryax) [reply pkg/stanza/adapter/config.go]: `initial_interval`, `max_interval` and `max_elapsed_time` have reasonable defaults and should be considered as secondary configurations while `enable` is the most important one. I users had to configure all the secondary configurations, I don't think there is a chance they miss `enabled`

File: internal/coreinternal/consumerretry/logs.go [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+				zap.Int("dropped_items", logs.LogRecordCount()),
+			)
+			return err
+		}
+
+		if errors.As(err, &retryableErr) {
+			logs = retryableErr.Data()
+		}
+
+		// TODO: take delay from the error once it is available in the consumererror package.
```
Comment (MovieStoreGuy) [inline internal/coreinternal/consumerretry/logs.go:95]: Should we have an issue to follow up on this? Think it would be a good first issue for anyone wanting to help.

File: internal/coreinternal/consumerretry/testutil.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+func NewMockLogsRejecter(acceptAfter int32) *MockLogsRejecter {
+	return &MockLogsRejecter{
+		acceptAfter: acceptAfter,
+		rejectCount: &atomic.Int32{},
+	}
+}
+
+func (m *MockLogsRejecter) ConsumeLogs(ctx context.Context, logs plog.Logs) error {
+	if m.acceptAfter < 0 {
+		return consumererror.NewPermanent(errors.New("permanent error"))
```
Comment (MovieStoreGuy) [inline internal/coreinternal/consumerretry/testutil.go:44]: Is it worth using a static error to make it easier to check with `errors.Is` ?

File: receiver/syslogreceiver/go.mod [hunk 8]
```diff
@@ -3,6 +3,7 @@ module github.com/open-telemetry/opentelemetry-collector-contrib/receiver/syslog
 go 1.19
 
 require (
+	github.com/open-telemetry/opentelemetry-collector-contrib/internal/coreinternal v0.0.0-00010101000000-000000000000
```
Comment (MovieStoreGuy) [inline receiver/syslogreceiver/go.mod:6]: Should we fudge the version number here or wait til next release for it to be captured there?

**Review by MovieStoreGuy — COMMENTED** [review]
Everything looks fine to me, just a couple questions.

File: internal/coreinternal/consumerretry/logs.go (hunk 6 above)
Comment (dmitryax) [reply internal/coreinternal/consumerretry/logs.go:95]: We have a draft PR already https://github.com/open-telemetry/opentelemetry-collector/pull/7439

File: internal/coreinternal/consumerretry/testutil.go (hunk 7 above)
Comment (dmitryax) [reply internal/coreinternal/consumerretry/testutil.go:44]: In this case `consumererror.IsPermanent(err)` should be used

File: receiver/syslogreceiver/go.mod (hunk 8 above)
Comment (dmitryax) [reply receiver/syslogreceiver/go.mod]: You're right. `coreinternal` is not a module created in this PR. I updated it

**Review by atoulme — APPROVED** [review]

File: pkg/stanza/adapter/config.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (djaglowski) [reply pkg/stanza/adapter/config.go]: This doesn't need to hold up the PR, but it seems like we should have a standard for when we need an `enabled` parameter.

Why do we use it in some places and not others? If it's better here, it seems we should be using it everywhere, unless there is a heuristic. I've added an item to the next SIG because I would like to understand this.

**Review by djaglowski — APPROVED** [review]

**Review by MovieStoreGuy — APPROVED** [review]

File: internal/coreinternal/consumerretry/logs.go [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+			return err
+		}
+
+		backoffDelayStr := backoffDelay.String()
+		span.AddEvent(
+			"Exporting failed. Will retry the request after interval.",
+			trace.WithAttributes(
+				attribute.String("interval", backoffDelayStr),
+				attribute.String("error", err.Error())))
+		strVal, _ := logs.ResourceLogs().At(0).Resource().Attributes().Get("com.splunk.source")
```
[the remark this answers is not in the document]
Comment (atoulme) [reply internal/coreinternal/consumerretry/logs.go:109]: I think this might have been used for debugging and can be removed

File: internal/coreinternal/consumerretry/logs.go [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+		lc.logger.Info(
+			"ConsumeLogs() failed. Will retry the request after interval.",
+			zap.Error(err),
+			zap.String("interval", backoffDelayStr),
+			zap.Int("logs_count", logs.LogRecordCount()),
+			zap.String("com.splunk.source", strVal.Str()),
```
Comment (atoulme) [inline internal/coreinternal/consumerretry/logs.go:115]:
```suggestion
```

File: internal/coreinternal/consumerretry/logs.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (atoulme) [reply internal/coreinternal/consumerretry/logs.go:109]:
```suggestion
```

[the remark this answers is not in the document]
Comment (dmitryax) [reply internal/coreinternal/consumerretry/logs.go:109]: Debugging stuff. Sorry

**Comment by tigrannajaryan** [comment]
Nice. If https://github.com/open-telemetry/opentelemetry-collector/pull/7516 gets accepted let's use to test the filelog receiver.
